### PR TITLE
feat: added beta factor support in VAE CV

### PR DESCRIPTION
**Pull request by PabloNA97**
I was trying to use the VAE CV as it is on my dataset (features from an MD simulation). I quickly noticed that my results were not as nice as in the example in the docs. Even using the same activation function and a small decoder as in the example (mlcolvar/docs/notebooks/tutorials/cvs_Autoencoder.ipynb). I was just observing a "blob" of points without any separation of metastable states:

image

See PCA on the same data for comparison:

image

With a bit of help from my favorite AI buddy (I'm not an expert on Deep learning or VAEs) I believe I identified the main issue. The regularization term in the loss together with my choice of hyperparameters and arquitecture were collapsing the latent space to the prior P(z) ~ N(0,I). So the model learned to reduce the loss just by reducing the KL term and mostly ignored the reconstruction term:

image

Before adding the beta factor I tried different arquitecture and hyperparameter choices. What improved the situation in the end was adding a low beta factor (0.001 in this case):

image

From what I understand, introducing a  beta factor to reduce the KL divergence term in the loss function can make the VAE training more robust to the architecture and hyper-parameter choices. Now I'm trying to use a Lightning callback during training to increase the beta factor after some epochs (see below):

```
import lightning

class KLAAnnealing(lightning.Callback):
    """
    Callback to anneal the KL divergence weight (beta).
    
    This callback linearly increases the beta value from 0 to max_beta
    over a specified number of epochs, starting from a given epoch.
    
    Inputs
    ------
    
    max_beta: 
        The final beta value to reach at the end of annealing.
    start_epoch: 
        The epoch at which to start annealing.
    n_epochs_anneal: 
        The number of epochs over which to anneal beta.
    """
    def __init__(self, max_beta=1.0, start_epoch=0, n_epochs_anneal=100):
        super().__init__()
        # The final beta value you want to reach
        self.max_beta = max_beta
        self.start_epoch = start_epoch
        self.n_epochs_anneal = n_epochs_anneal

    def on_train_epoch_start(self, trainer, pl_module):
        current_epoch = trainer.current_epoch
        if current_epoch < self.start_epoch:
            beta = 0.0
        elif current_epoch >= self.start_epoch + self.n_epochs_anneal:
            beta = self.max_beta
        else:
            # Linearly increase beta
            progress = (current_epoch - self.start_epoch) / self.n_epochs_anneal
            beta = self.max_beta * progress
            
        # Set beta in the LightningModule
        # Assumes your pytorch_lightning module has a beta attribute
        if not hasattr(pl_module, 'beta'):
            logger.warning("The LightningModule does not have a 'beta' attribute. "
                           "Please ensure it is defined to use KLAAnnealing.")
            return
        pl_module.beta = beta
        pl_module.log('beta', beta, on_step=False, on_epoch=True)
```
One has to add this callback to the Trainer and beta factor will be modified accordingly:

image

This beta factor can also be used to obtain disentangled components in the latent space by collapsing the uninformative ones.  See [1]

[1] Burgess, Christopher P., Irina Higgins, Arka Pal, Loic Matthey, Nick Watters, Guillaume Desjardins, and Alexander Lerchner. “Understanding Disentangling in $β$-VAE.” arXiv, April 10, 2018. https://doi.org/10.48550/arXiv.1804.03599.

### CODE REVIEW

**Comment by luigibonati** [comment]
Hi @PabloNA97,
thanks a lot for your detailed report and contribution!

You’re right: we implemented the VAE mainly to provide a basic building block for exploring methods from the literature, but we never really tested it thoroughly as a collective variable. As you found out, if the KL loss dominates, the latent space tends to collapse toward the prior $P(z) \sim \mathcal{N}(0, I)$, which is far from ideal when dealing with metastability. So implementing also the beta-VAE seems a nice tool to have.

If you’re interested in contributing a small tutorial or example notebook based on your findings, it would be very welcome!
Luigi

**Comment by PabloNA97** [comment]
Hi @luigibonati,

I would gladly contribute with a small tutorial. If I have the time to put this together I will add it to a new PR and share it :) 

Thanks!

Pablo

File: mlcolvar/core/loss/elbo.py [hunk 1]
```diff
@@ -57,6 +59,17 @@ def forward(
...
             Shape ``(n_batches, latent_features)``. The logarithm of the variances
             of the Gaussian distributions associated to the inputs.
+        beta : float, optional
+            A scaling factor for the KL divergence term. The default is 1.0,
+            which means that the KL divergence is not scaled. If set to a value
+            greater than 1, it will increase the weight of the KL divergence
+            term in the loss function (useful to increase regularization). 
+            If set to a value less than 1, it will decrease the weight
+            of the KL divergence term (useful to avoid posterior collapse)
+        decompose : bool, optional
```
[the remark this answers is not in the document]
Comment (PabloNA97) [reply mlcolvar/core/loss/elbo.py]: Done :)

**Comment by EnricoTrizio** [comment]
Ciao @PabloNA97 , in #207 I updated the codecov action so that it can also cover external PRs to have an idea of the test coverage. Do you mind either pulling the changes from the main branch of the library or giving access to maintainers to your fork?
This way we can check everything is covered and close the PR.
Thankss!
